lib/plugins/aws/package/compile/events/stream/index.js

```diff
@@ -11,40 +11,83 @@ class AwsCompileStreamEvents {
       'package:compileEvents': this.compileStreamEvents.bind(this),
     };
 
-    // TODO: Complete schema, see https://github.com/serverless/serverless/issues/8034
     this.serverless.configSchemaHandler.defineFunctionEvent('aws', 'stream', {
-      anyOf: [{ type: 'string' }, { type: 'object' }],
+      anyOf: [
+        { $ref: '#/definitions/awsArnString' },
+        {
+          type: 'object',
+          properties: {
+            arn: { $ref: '#/definitions/awsArn' },
+            type: { enum: ['dynamodb', 'kinesis'] },
+            batchSize: { type: 'integer', minimum: 1, maximum: 10000 },
+            parallelizationFactor: { type: 'integer', minimum: 1, maximum: 10 },
+            startingPosition: { enum: ['LATEST', 'TRIM_HORIZON'] },
+            enabled: { type: 'boolean' },
+            consumer: { oneOf: [{ const: true }, { $ref: '#/definitions/awsArn' }] },
+            batchWindow: { type: 'integer', minimum: 0, maximum: 300 },
+            maximumRetryAttempts: { type: 'integer', minimum: -1, maximum: 10000 },
+            bisectBatchOnFunctionError: { type: 'boolean' },
+            maximumRecordAgeInSeconds: {
+              oneOf: [
+                { type: 'integer', minimum: -1, maximum: -1 },
+                { type: 'integer', minimum: 60, maximum: 604800 },
+              ],
+            },
+            destinations: {
+              type: 'object',
+              properties: {
+                onFailure: {
+                  anyOf: [
+                    { $ref: '#/definitions/awsArnString' },
+                    {
+                      type: 'object',
+                      properties: {
+                        arn: { $ref: '#/definitions/awsArn' },
+                        type: { enum: ['sns', 'sqs'] },
+                      },
+                      additionalProperties: false,
+                      required: ['arn'],
+                      oneOf: [
+                        {
+                          properties: {
+                            arn: { $ref: '#/definitions/awsCfFunction' },
+                          },
+                          required: ['type'],
+                        },
+                        {
+                          properties: {
+                            arn: { $ref: '#/definitions/awsArnString' },
+                          },
+                        },
+                      ],
```

Comment:
I wonder wouldn't it be more clear (less confusing), to totally discard `properties` above, and simply configure both variants fully in  scope `oneOf`.

In current form it's a bit difficult to read. We have three `properties` references that outline two acceptable forms.

Reply:
agreed

Reply:
@medikoo one small problem to that :
- if I remove completly arn definition from `properties` base field, I get a validation error `'functions.hello.events[0].stream': unrecognized property 'arn'` since `additionalProperties: false` is set.
- if I want this error to not be thrown, `arn` needs to be defined outside of the `oneOf` clause. All statements in `oneOf` are complementary validation logic : both base `object` JSON schema and one statement of `oneOf` keyword must validate.

I have then 3 solutions :

- keeping the PR as it is now where `oneOf` clauses about `arn` definitions are subparts of  base `properties.arn` definition.
- specifying `arn: {}` definition in the base `properties` keyword  in order to tell that there is an `arn` property that will be defined later. The risk is someone opening up to widely the `oneOf` clauses about arn and have invalid arn specified. (what I updated the PR to with the last commit, for me this would be the preferred way since there are very small chances above scenario actually happen)
- removing `additionalProperties: false` clause

WDYT ?

Reply:
Yes, I also think `arn: {}` would be the way to go. So in `properties` we indicate purely that  it's a recognized property, but full schema is outlined in `oneOf`.

It gives clear separation, and minimal risk for mismatch error (e.g. there could be configuration issue, where in `properties` we configure schema that doesn't align with one configured in `oneOf` options)